zebra-network/src/peer_set/set.rs

```diff
@@ -447,14 +475,23 @@ where
     /// Adds a busy service to the unready list,
     /// and adds a cancel handle for the service's current request.
     fn push_unready(&mut self, key: D::Key, svc: D::Service) {
+        let peer_version = svc.version();
         let (tx, rx) = oneshot::channel();
-        self.cancel_handles.insert(key, tx);
+
         self.unready_services.push(UnreadyService {
             key: Some(key),
             service: Some(svc),
             cancel: rx,
             _req: PhantomData,
         });
+
+        if peer_version >= self.minimum_peer_version.current() {
+            self.cancel_handles.insert(key, tx);
+        } else {
+            // Cancel any request made to the service because it is using an outdated protocol
+            // version.
+            std::mem::drop(tx);
```

Comment:
Likely Bug: `UnreadyService::poll` doesn't check for `Err(oneshot::Canceled)`. But it probably should.

https://docs.rs/futures/0.3.18/futures/channel/oneshot/struct.Canceled.html

https://github.com/ZcashFoundation/zebra/blob/ebc40f8af7a6ca0a375d2effe728a9fc79a64b96/zebra-network/src/peer_set/unready_service.rs#L40

Reply:
Oops, I totally missed this. I now made it send a `CancelClientWork` message instead, because it fixes the issue and makes the intent clearer (and is consistent with other places). I opened an issue (#3130) to try to improve the `UnreadyService` handling of the cancel handle to avoid any issues in the future.